node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+	{
+		// startup: register the network protocol with the bridge.
+		ctx.send_message(AllMessages::NetworkBridge(
+			NetworkBridgeMessage::RegisterEventProducer(Self::PROTOCOL_ID, network_update_message),
+		))
+		.await?;
+
+		// work: process incoming messages from the overseer and process accordingly.
+		let mut tracker = Tracker::default();
+		loop {
+			{
+				let message = ctx.recv().await?;
+				match message {
+					FromOverseer::Communication { msg } => {
+						// another subsystem created this signed availability bitfield messages
+						match msg {
+							// relay_message a bitfield via gossip to other validators
+							BitfieldDistributionMessage::DistributeBitfield(
+								hash,
+								signed_availability,
+							) => {
+								trace!(target: "bitd", "Processing DistributeBitfield");
+								let job_data = &mut tracker.per_relay_parent.get_mut(&hash).expect("Overseer does not send work items related to relay parents that are not part of our workset. qed");
+
+								let validator = {
+									job_data
+										.validator_set
+										.get(signed_availability.validator_index() as usize)
+										.expect("Our own validation index exists. qed")
+								}
+								.clone();
+
+								let peer_views = &mut tracker.peer_views;
+								let msg = BitfieldGossipMessage {
+									relay_parent: hash,
+									signed_availability,
+								};
+
+								relay_message(&mut ctx, job_data, peer_views, validator, msg)
+									.await?;
+							}
+							BitfieldDistributionMessage::NetworkBridgeUpdate(event) => {
+								trace!(target: "bitd", "Processing NetworkMessage");
+								// a network message was received
+								if let Err(e) =
+									handle_network_msg(&mut ctx, &mut tracker, event).await
+								{
+									warn!(target: "bitd", "Failed to handle incomming network messages: {:?}", e);
+								}
+							}
+						}
+					}
+					FromOverseer::Signal(OverseerSignal::StartWork(relay_parent)) => {
+						trace!(target: "bitd", "Start {:?}", relay_parent);
+						// query basic system parameters once
+						// @todo assumption: these cannot change within a session
+						let (validator_set, signing_context) =
+							query_basics(&mut ctx, relay_parent).await?;
+
+						let _ = tracker.per_relay_parent.insert(
+							relay_parent,
+							PerRelayParentData {
+								signing_context,
+								validator_set,
+								..Default::default()
+							},
+						);
+					}
+					FromOverseer::Signal(OverseerSignal::StopWork(relay_parent)) => {
+						trace!(target: "bitd", "Stop {:?}", relay_parent);
+						// @todo assumption: it is good enough to prevent additional work from being
+						// scheduled, the individual futures are supposedly completed quickly
+						let _ = tracker.per_relay_parent.remove(&relay_parent);
+					}
+					FromOverseer::Signal(OverseerSignal::Conclude) => {
+						trace!(target: "bitd", "Conclude");
+						tracker.per_relay_parent.clear();
```

Comment:
This is redundant, since exiting the function with `return` `Drop`s the tracker.